untrusted/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+// See https://download.01.org/intel-sgx/sgx-dcap/1.13/linux/docs/Intel_SGX_Enclave_Common_Loader_API_Reference.pdf
+//
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+use std::{ffi::CString, mem::MaybeUninit, os::raw::c_int};
+
+include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+
+#[derive(Default)]
+pub struct Enclave {
+    // The filename of the enclave
+    filename: CString,
+
+    // The enclave ID, assigned by the sgx interface
+    // Will be `None` when the enclave has not been created.
+    id: Option<sgx_enclave_id_t>,
+
+    // `true` if the enclave should be created in debug mode
+    debug: bool,
+}
+
+impl Enclave {
+    /// Returns an Enclave for the provided signed enclave.
+    ///
+    /// # Arguments
+    ///
+    /// * `filename` - The name of the enclave file.  This should be a signed
+    ///     enclave.
+    pub fn new(filename: &str) -> Enclave {
+        let filename = CString::new(filename).expect("Can't convert enclave filename to CString.");
+        Enclave {
+            filename,
+            ..Default::default()
+        }
+    }
+
+    /// Toggle debugging of the enclave on or off.  The default is off.
+    ///
+    /// # Arguments
+    ///
+    /// * `debug` - `true` to enable enclave debugging, `false` to disable it.
+    pub fn debug(&mut self, debug: bool) -> &mut Enclave {
+        self.debug = debug;
+        self
+    }
+
+    /// Get the ID for this instance.  The ID will not be valid to use in SGX
+    /// calls once this instance has dropped.
+    ///
+    /// The return value is intentionally a reference to a `sgx_enclave_id_t`.
+    /// This allows consumers to leverage the lifetime of the [Enclave]
+    /// instance, preventing the call of SGX functions after the [Enclave] has
+    /// been dropped.
+    pub fn get_id(&self) -> Option<&sgx_enclave_id_t> {
+        Option::from(&self.id)
+    }
+
+    /// Create the enclave
+    ///
+    /// Will talk to the SGX SDK to create the enclave.  Once the enclave has
+    /// been created then calls on the enclave can be made.  See
+    /// [Enclave::get_id()]
+    ///
+    /// # Returns
+    ///
+    /// [_status_t_SGX_SUCCESS] when the enclave is created successfully.
+    ///
+    /// See
+    /// <https://download.01.org/intel-sgx/sgx-dcap/1.13/linux/docs/Intel_SGX_Enclave_Common_Loader_API_Reference.pdf>
+    /// for error codes and their meaning.
+    pub fn create(&mut self) -> sgx_status_t {
+        let mut launch_token: sgx_launch_token_t = [0; 1024];
```

Comment:
good catch!

Seems like I should update the doc comment to say this function should only every be called once per instance. 
I can't think of a term for a function like that, "re-entrant" is close but slightly different as it usually implies calling the function more than once concurrently.

While calling this function more than once is a slightly different error case than failing to create the enclave from the C API, I'm not sure the rust API should have different ways to fail, that seems inconsistent for an eventual  end user.
Perhaps for now this should return something like `_status_t_SGX_ERROR_UNEXPECTED` (I don't like that error code, but...).   Then longer term or as a follow on this should return a `Result` such that `_status_t_SGX_SUCCESS` => Ok(). and the other values come back as an `Error` enum.  Then an error like `InUse` could be returned for this attempted recreation logic.

Reply:
One way to implement this is with a Builder pattern. Then you can have a method that consumes the instance, e.g. `fn create(self) -> Enclave`

Reply:
I don't think I fully follow. Are you saying to return itself in the result
```
impl Enclave {
    pub fn create(self) -> Result<Enclave, Error>
```
or are you thinking of having 2 types
```
impl Builder {
    pub fn create(self) -> Result<Enclave, Error>

Reply:
The latter

Reply:
Yeah, in both cases, all the `Enclave::create()` code should go into `Enclave::new()`, the only difference is whether `Enclave::new()` is public or not.

Reply:
Builder pattern is now  implemented.

> all the Enclave::create() code should go into Enclave::new()
I didn't do this and left the logic in `create()` since it could fail.  Doing a quick scrape on the googles for `new()` it seems like most return the item not a `Result`